### PR TITLE
Changes for iOS simulator runs

### LINKED ISSUE #1227: Add ios simulator support for p2p tests
The p2p tests currently cannot be run on ios device or simulator. We should add support to run p2p tests on ios simulator as a first step.

### PR DESCRIPTION
**Pull request by raghusarangapani**
#### Fixes #1227.

- [x] Ran `flake8`
- [x] Ran `run_repo_tests.sh`

#### Changes proposed in this pull request:
- Run p2p tests using ios simulator

### CODE REVIEW

File: keywords/LiteServiOS.py [hunk 1]
```diff
@@ -51,13 +87,61 @@ def install(self):
...
+        package_name = "LiteServ-iOS.app"
+        app_dir = "LiteServ-iOS"
+
+        if self.storage_engine == "SQLCipher":
+            package_name = "LiteServ-iOS-SQLCipher.app"
+            app_dir = "LiteServ-iOS-SQLCipher"
+
+        app_path = "{}/{}/{}".format(BINARY_DIR, app_dir, package_name)
+        output = subprocess.check_output([
+            "ios-sim", "--devicetypeid", device, "start"
```
Comment (sethrosetter) [inline keywords/LiteServiOS.py]: What happens if simulator is started already?

File: keywords/LiteServiOS.py [hunk 2]
```diff
@@ -51,13 +87,61 @@ def install(self):
...
+
+        log_info(output)
+
+        # Wait for the device to boot up
+        start = time.time()
+        while True:
+            if time.time() - start > CLIENT_REQUEST_TIMEOUT:
+                raise LiteServError("iPhone Simulator failed to start")
+
+            output = subprocess.Popen([
```
Comment (sethrosetter) [inline keywords/LiteServiOS.py]: Can you add a comment to highlight what you are doing here?

File: keywords/LiteServiOS.py [hunk 3]
```diff
@@ -51,13 +87,61 @@ def install(self):
...
+            status = output.split(" ")[2]
+            if status == "A":
+                log_info("iPhone Simulator seems to have booted up")
+                break
+            else:
+                log_info("Waiting for the iPhone Simulator to boot up")
+                time.sleep(1)
+                continue
+
+    def remove_device(self):
```
Comment (sethrosetter) [inline keywords/LiteServiOS.py]: Why are you touching remove device if we are only working with simulator?

File: keywords/LiteServiOS.py [hunk 4]
```diff
@@ -70,6 +154,44 @@ def remove(self):
...
+        if self.storage_engine == "SQLCipher":
+            bundle_id = "com.couchbase.LiteServ-iOS-SQLCipher"
+
+        log_info("Removing LiteServ")
+
+        self.stop()
+
+        # Get the device ID
+        device_id = None
+        list_output = subprocess.Popen(["xcrun", "simctl", "list"], stdout=subprocess.PIPE)
```
Comment (sethrosetter) [inline keywords/LiteServiOS.py]: What if multiple simulators are running for multiple test runs?

File: keywords/LiteServBase.py [hunk 5]
```diff
@@ -52,8 +52,11 @@ def _verify_not_running(self):
         log_r(resp)
         raise LiteServError("There should be no service running on the port")
 
-    def _wait_until_reachable(self):
-        url = "http://{}:{}".format(self.host, self.port)
+    def _wait_until_reachable(self, port=None):
+        if not port:
```
Comment (raghusarangapani) [inline keywords/LiteServBase.py]: The LiteServ iOS starts off with the default admin port 59850 and public port 49850, which is not the same as the --liteserv-port we specify on the command line. Switching to the new port specified on the command line happens through a REST API call to /start. The _wait_until_reachable takes the port specified on the command line to check for the status as soon as the LiteServ is launched. This will obviously time out as the /start REST API call happens much later, when the tests actually run. This change enables checking the status with any port. The default behavior(self.port) is not changed.

File: keywords/LiteServiOS.py (hunk 1 above)
Comment (raghusarangapani) [inline keywords/LiteServiOS.py]: No error is raised. It will go ahead and install the .app

File: keywords/LiteServiOS.py (hunk 2 above)
Comment (raghusarangapani) [inline keywords/LiteServiOS.py]: Done

File: keywords/LiteServiOS.py (hunk 4 above)
Comment (raghusarangapani) [inline keywords/LiteServiOS.py]: I was not able to successfully launch multiple simulators, though some articles on the internet mentioned it is possible.

File: keywords/LiteServiOS.py (hunk 3 above)
Comment (raghusarangapani) [inline keywords/LiteServiOS.py]: Removed the 2 line change

File: keywords/LiteServiOS.py [hunk 6]
```diff
@@ -79,25 +204,56 @@ def start(self, logfile_name):
...
+            db_flags = []
+            for db_name in REGISTERED_CLIENT_DBS:
+                db_flags.append("{}:pass".format(db_name))
+            db_flags = ",".join(db_flags)
+
+            log_info("Running with db_flags: {}".format(db_flags))
+            data["dbpasswords"] = db_flags
+
+        self._wait_until_reachable(port=59850)
+        liteserv_admin_url = "http://{}:59850".format(self.host)
```
Comment (sethrosetter) [inline keywords/LiteServiOS.py]: Since this is used multiple times, can you store as a class property?

File: keywords/LiteServiOS.py [hunk 7]
```diff
@@ -131,8 +287,20 @@ def stop(self):
 
         liteserv_admin_url = "http://{}:59850".format(self.host)
         log_info("Stopping LiteServ: {}".format(liteserv_admin_url))
-        resp = self.session.put("{}/stop".format(liteserv_admin_url))
-        log_r(resp)
-        resp.raise_for_status()
+
+        count = 0
+        while count < MAX_RETRIES:
```
Comment (sethrosetter) [inline keywords/LiteServiOS.py]: Why are you retrying here? `liteserv.stop` should only be called if a LiteServ is running.

File: keywords/LiteServiOS.py (hunk 6 above)
Comment (raghusarangapani) [inline keywords/LiteServiOS.py]: I thought that this was specific to iOS, I don't see such a usage anywhere in LiteServAndroid.py.

File: keywords/LiteServiOS.py (hunk 7 above)
Comment (raghusarangapani) [inline keywords/LiteServiOS.py]: Good catch, this was left over code from my experiments.

File: keywords/LiteServiOS.py (hunk 6 above)
Comment (sethrosetter) [inline keywords/LiteServiOS.py]: it is iOS specific. Store it as a `LiteServiOS.py` property

File: keywords/LiteServiOS.py [hunk 8]
```diff
@@ -137,6 +144,17 @@ def install(self):
                 time.sleep(1)
                 continue
 
+        # Get the device ID
+        list_output = subprocess.Popen(["xcrun", "simctl", "list"], stdout=subprocess.PIPE)
```
Comment (sethrosetter) [inline keywords/LiteServiOS.py]: why not use `subprocess.check_ouput()` here?

File: keywords/LiteServiOS.py [hunk 9]
```diff
@@ -137,6 +144,17 @@ def install(self):
...
+        output = subprocess.check_output(('grep', 'Booted'), stdin=list_output.stdout)
+
+        for line in output.splitlines():
+            if "Phone" in line:
+                self.device_id = re.sub(' +', ' ', line).strip()
+                self.device_id = self.device_id.split(" ")[4]
+                self.device_id = self.device_id.strip('(')
+                self.device_id = self.device_id.strip(')')
```
Comment (sethrosetter) [inline keywords/LiteServiOS.py]: Should we raise an exception is the device id is still `None` at this point?

File: keywords/LiteServiOS.py [hunk 10]
```diff
@@ -282,12 +290,18 @@ def stop(self):
...
         log_r(resp)
         resp.raise_for_status()
 
+        # Using --exit in ios-sim means, --log has no effect
+        # Have to separately copy the simulator logs
+        if self.logfile_name and self.device_id:
+            home = os.environ['HOME']
+            ios_log_file = "{}/Library/Logs/CoreSimulator/{}/system.log".format(home, self.device_id)
+            copyfile(ios_log_file, self.logfile_name)
+            open(ios_log_file, 'w').close()
```
Comment (sethrosetter) [inline keywords/LiteServiOS.py]: Why is this needed?

File: keywords/LiteServiOS.py (hunk 8 above)
Comment (raghusarangapani) [inline keywords/LiteServiOS.py]: I am using check_output in the next line to grep Booted. From my research, this is equivalent to bash's pipe feature.

File: keywords/LiteServiOS.py (hunk 9 above)
Comment (raghusarangapani) [inline keywords/LiteServiOS.py]: Done

File: keywords/LiteServiOS.py (hunk 10 above)
Comment (raghusarangapani) [inline keywords/LiteServiOS.py]: Are the comments insufficient?

Comment (sethrosetter) [inline keywords/LiteServiOS.py]: Yes. Looks good.

File: keywords/LiteServiOS.py (hunk 9 above)
Comment (sethrosetter) [inline keywords/LiteServiOS.py]: Thanks

File: keywords/LiteServiOS.py [hunk 11]
```diff
@@ -1,39 +1,80 @@
...
+from keywords.constants import REGISTERED_CLIENT_DBS
+from keywords.constants import CLIENT_REQUEST_TIMEOUT
 
 
 class LiteServiOS(LiteServBase):
 
     def __init__(self, version_build, host, port, storage_engine):
-
-        # Initialize baseclass properies
         super(LiteServiOS, self).__init__(version_build, host, port, storage_engine)
```
Comment (sethrosetter) [inline keywords/LiteServiOS.py]: add a check for `forestdb` and `forestdb+encryption`. If the user tries to use these with iOS, raise an exception.

Comment (raghusarangapani) [inline keywords/LiteServiOS.py]: Done
